### DISCUSSION TITLE
Defining signals as vectors

### DISCUSSION THREAD

**Discussion by corbinklett** [python-control/python-control#1008] [Ideas]
It seems like signals can only be defined by giving an integer count (in which case the signals are named with 'u', 'y', or 'x'), or by naming individual vector components. Is it possible to specify vector signals, like:

`airplane = ct.nlsys(updfcn=dynamics, outfcn=None, inputs=('forces[3]', 'moments[3]'), outputs=...`

Would it be difficult for me to add this feature?

From the documentation:
inputs (int, list of str or None, optional) – Description of the system inputs. This can be given as an integer count or as a list of strings that name the individual signals. If an integer count is specified, the names of the signal will be of the form s[i] (where s is one of u, y, or x)

**Comment by corbinklett** [comment]
Another example of how vector signals would be used, in the outputs naming:

```python
forces_moments = ct.interconnect(
    (aerodynamics, propulsion, gravity, sum_forces_moments),
    name="forces_moments",
    connections=[
        # input <- output
        ['sum_forces_moments.u[:6]', 'aerodynamics.y'],
        ['sum_forces_moments.u[6:12]', 'propulsion.y'],
        ['sum_forces_moments.u[12:15]', 'gravity.y'],
    ],
    inplist = ['aerodynamics.coef'],
    outlist = ['sum_forces_moments.y', 'aerodynamics.y[:3]'],
    outputs = ['F_total[3]', 'M_total[3]', 'F_aero[3]']
)
```

**Comment by murrayrm** [comment]
It would be great to get some additional processing of vector signals.  A few notes on the above:
* <s>For the connections keyword, the syntax is `[output, input1, input2, ...]` so you would want ['aerodynamics.y', 'sum_forces_moments.u[:6]'], I think.</s>
* I'm not sure that 'F_total[3]' is the best syntax to use, since that also refers to the fourth item in the signal 'F_total'.  Probably that would be better as 'F_total[0:3]'?  So perhaps just handle the slice notation?
* There is a set of `_prefix` keywords that can accomplish some of this, though your proposed syntax is easier.

**Reply by corbinklett** [reply]
The idea of F_total[3] followed by M_total[3] was to map the first three elements of sum_forces_moments.y to F_total, then the next three to M_total. Will keep thinking about vector signal assignments as my project gets bigger.

It seems like the input comes first when using the connections keyword. If you run the script in https://github.com/python-control/python-control/issues/1009 but switch the connection order, you get `ValueError: couldn't find input signal 'input_sys.y'`

**Comment by corbinklett** [comment]
Adding a note to also think about the case when one input to an interconnected system needs to be accessed by multiple subsystems. For example, what if each of the aerodynamics, propulsion, and gravity models needs access to the vehicle rigid body state vector? Currently, the state vector would have to be connected to the interconnected system 3 separate times (which could be a list 36+ items long)

**Reply by murrayrm** [reply]
What I see a lot of people doing, and what I typically do when using python-control for classes, is to name the signals in a way that allows automatic interconnection (no `connections` keyword) do its thing.  In that case you can just have the aerodynamics, propulsion, and gravity models all have inputs with the appropriate names and the vehicle rigid body model use the same names for its outputs, and everything will get connected up.

List comprehensions can also be helpful, if the names are all the same.  So something like:
```
connections = [[f'{subsys}.u[0:3]', 'vehicle.y[0:3]'] for subsys in ['aero', 'prop', 'grav']]
```
could work.

**Reply by corbinklett** [reply]
That tip about automatic connections was a game changer for writing a large sim with simple code. Thanks!
